### ISSUE TITLE
What the temperature in softmax layer?

### ISSUE THREAD

**Issue by norris9410** [lshug/first-order-model-tf#9]
Hi Levan, thank you very much for making it into tf version!! Can I ask one question if you don't mind? 

So when I was reading both codes from yours and the from the original paper, I notice that a term "temperature" in involved in softmax layer. For example, in your "firstordermodel.py" line 99,

`out = keras.activations.softmax(out / self.temperature, axis=1)`

could you tell me what is the temperature mean? I tried to look at the origin paper but still didn't find any explanation. Is there anything I missed?

look forward to your replies.

Norris

**Comment by lshug** [comment]
Hello.

At the end of the kp detector, representation obtained from the convolution layers contains num_kp numbers at each spatial point. The numbers are passed through a sigmoid one of the spatial axes (not the num_kp axis), which results in there being an approximately gaussian distribution of numbers in [0,1] range across the image for each kp. Each such slice is interpreted as a heatmap that shows the probability of a point being a good location for the kp at each spatial point. 

The point of dividing the representation by the temperature pre-sigmoid is to "heat up the heatmap": it increases the absolute value of the numbers, and thus makes more of the numbers align towards the extreme ends of the sigmoid's range, i.e. it ends up with more numbers being closer to 1 and 0 and less numbers being somewhere in-between.

The reason this is done is explained by what comes next. The point with the highest predicted probability can't just be picked from the spatial map, because "get the coordinates of the point with the highest number" is not a differentiable function and the kp detector as a whole needs to be differentiable to be optimizable by gradient descent. Because of this, the architecture uses the differentiable argmax product thingie introduced in [Robinson et al., 2019.](https://arxiv.org/pdf/1903.11633.pdf): the result of the softmax at each point is multiplied by that point's x-y coordinates, and the resultant products are summed and turned into an int. Getting why this works is easier with a 1d representation:

```
x:    0      1      2      3      4      | sum of products
p:   0.10   0.07   0.03   0.30   0.50
pr:  0.00   0.07   0.06   0.90   2.00=       3.03
```
The final result is floor(3.03)=3, which is just about where the most probability mass is concentrated. In the end, you're just averaging out the pixel coordinates weighted by the predicted probabilities.

Without scaling pre-sigmoid representations by temperature (represented as β in Robinson), a lot of the probability values could fall somewhere closely around 1/h, and that would throw off the averaging by moving the final result to somewhere in-between the extremes, when we're really only interested in the extremes. By applying the temperature we're making higher numbers go higher and lower numbers go lower.


One nuance to be considered is that this model doesn't follow Robinson et al. equation completely: softmaxing is only done on one of the spatial dimensions, so the row sums results in a column-sized vector of ones, instead of the whole spatial sum summing up to a single 1. This is done for practical purposes: neither pytorch's F.softmax nor tf.reduce_max used inside tf.keras.activations.softmax support softmaxing along multiple axes. In practice you end up with mostly the same result though.

**Comment by norris9410** [comment]
Cannot find a better explanation than this. Thank you very much!
